server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -562,15 +643,39 @@ static void registerNodeSearchAction(
         final int searchPoolMax = threadPool.info(ThreadPool.Names.SEARCH).getMax();
         transportService.registerRequestHandler(
             NODE_SEARCH_ACTION_NAME,
-            EsExecutors.DIRECT_EXECUTOR_SERVICE,
+            threadPool.executor(ThreadPool.Names.SEARCH_COORDINATION),
             NodeQueryRequest::new,
             (request, channel, task) -> {
-                final CancellableTask cancellableTask = (CancellableTask) task;
+                final SearchRequest searchRequest = request.searchRequest;
+                ShardSearchRequest[] shardSearchRequests = null;
+                IntUnaryOperator shards = IntUnaryOperator.identity();
                 final int shardCount = request.shards.size();
-                int workers = Math.min(request.searchRequest.getMaxConcurrentShardRequests(), Math.min(shardCount, searchPoolMax));
+                if (shardCount > 1 && hasPrimaryFieldSort(searchRequest.source())) {
+                    try {
+                        shardSearchRequests = new ShardSearchRequest[shardCount];
+                        @SuppressWarnings("rawtypes")
+                        final MinAndMax[] minAndMax = new MinAndMax[shardCount];
+                        for (int i = 0; i < minAndMax.length; i++) {
+                            ShardSearchRequest r = buildShardSearchRequestForLocal(request, request.shards.get(i));
+                            shardSearchRequests[i] = r;
+                            var canMatch = searchService.canMatch(r);
+                            if (canMatch.canMatch()) {
+                                r.setRunCanMatchInQueryPhase(false);
```

Comment:
A bit hacky for now and we should clean this logic up further, but somewhat of a nice speedup to save one round of creating a searcher and whatnot and a win we should take right away IMO.